File: python_src/tools/generate_certificates.py
from __future__ import annotations

from pathlib import Path
from typing import Any

from .common import run_command, text_response, ToolResponse, require_ident


def generate_certificates(args: dict[str, Any]) -> ToolResponse:
    route = args.get("route")
    operation = args.get("operation", "generate")
    working_directory = args.get("workingDirectory") or str(Path.cwd())

    try:
        namespace = require_ident(args.get("namespace"), "namespace")
        release_name = require_ident(args.get("releaseName"), "releaseName")
        if route:
            route = require_ident(route, "route")

        if operation == "verify":
            return _verify_certificates(namespace, release_name, working_directory)

        command = ["./generate-certificates.sh", namespace, release_name]
        if route:
            command.append(route)

        stdout, stderr = run_command(command, cwd=working_directory, max_bytes=5 * 1024 * 1024)

        response = "# Certificate Generation\n\n"
        response += f"**Namespace:** {namespace}\n"
        response += f"**Release Name:** {release_name}\n"
        if route:
            response += f"**Route:** {route}\n"
        response += f"**Operation:** {operation}\n\n"
        response += f"## Output\n\n```\n{stdout}```\n"

        if stderr:
            response += f"\n## Warnings\n\n```\n{stderr}```\n"

        response += "\n## Certificate Files Generated\n\n"
        response += f"- CA Certificate: `{namespace}-ca.crt`\n"
        response += f"- Server Certificate: `{namespace}-server.crt`\n"
        response += f"- Server Key: `{namespace}-server.key`\n"
        response += f"- Truststore: `{namespace}-truststore.jks`\n"

        return text_response(response)
    except Exception as error:  # noqa: BLE001
        return text_response(f"Certificate operation failed: {error}", is_error=True)


def _verify_certificates(namespace: str, release_name: str, working_dir: str) -> ToolResponse:
    try:
        work = Path(working_dir).resolve()
        cert_file = (work / f"{namespace}-server.crt").resolve()
        if work not in cert_file.parents:
            raise ValueError("Certificate path is outside the working directory")

        stdout, _ = run_command(
            ["openssl", "x509", "-in", str(cert_file), "-text", "-noout"],
            cwd=str(work),
        )

        response = "# Certificate Verification\n\n"
        response += f"**Namespace:** {namespace}\n"
        response += f"**Release Name:** {release_name}\n\n"
        response += f"## Certificate Details\n\n```\n{stdout}```\n"
        return text_response(response)
    except Exception:  # noqa: BLE001
        return text_response("Certificate verification failed. Certificates may not exist or are invalid.", is_error=True)

File: src/tools/generate-certificates.ts
import { exec } from "child_process";
import { promisify } from "util";
import * as path from "path";

const execAsync = promisify(exec);

interface GenerateCertificatesArgs {
  namespace: string;
  releaseName: string;
  route?: string;
  operation?: "generate" | "verify" | "regenerate";
  workingDirectory?: string;
}

export async function generateCertificates(args: any) {
  const {
    namespace,
    releaseName,
    route,
    operation = "generate",
    workingDirectory = process.cwd(),
  } = args as GenerateCertificatesArgs;

  try {
    let command = `cd "${workingDirectory}" && ./generate-certificates.sh ${namespace} ${releaseName}`;
    
    if (route) {
      command += ` ${route}`;
    }

    if (operation === "verify") {
      return await verifyCertificates(namespace, releaseName, workingDirectory);
    }

    // Execute certificate generation
    const { stdout, stderr } = await execAsync(command, {
      cwd: workingDirectory,
      maxBuffer: 1024 * 1024 * 5,
    });

    let response = `# Certificate Generation\n\n`;
    response += `**Namespace:** ${namespace}\n`;
    response += `**Release Name:** ${releaseName}\n`;
    if (route) response += `**Route:** ${route}\n`;
    response += `**Operation:** ${operation}\n\n`;
    
    response += `## Output\n\n\`\`\`\n${stdout}\`\`\`\n`;
    
    if (stderr) {
      response += `\n## Warnings\n\n\`\`\`\n${stderr}\`\`\`\n`;
    }

    response += `\n## Certificate Files Generated\n\n`;
    response += `- CA Certificate: \`${namespace}-ca.crt\`\n`;
    response += `- Server Certificate: \`${namespace}-server.crt\`\n`;
    response += `- Server Key: \`${namespace}-server.key\`\n`;
    response += `- Truststore: \`${namespace}-truststore.jks\`\n`;

    return {
      content: [{ type: "text", text: response }],
    };
  } catch (error) {
    const errorMessage = error instanceof Error ? error.message : String(error);
    return {
      content: [
        {
          type: "text",
          text: `Certificate operation failed: ${errorMessage}`,
        },
      ],
      isError: true,
    };
  }
}

async function verifyCertificates(
  namespace: string,
  releaseName: string,
  workingDir: string
) {
  try {
    const certFile = path.join(workingDir, `${namespace}-server.crt`);
    const command = `openssl x509 -in "${certFile}" -text -noout`;

    const { stdout } = await execAsync(command, { cwd: workingDir });

    let response = `# Certificate Verification\n\n`;
    response += `**Namespace:** ${namespace}\n`;
    response += `**Release Name:** ${releaseName}\n\n`;
    response += `## Certificate Details\n\n\`\`\`\n${stdout}\`\`\`\n`;

    return {
      content: [{ type: "text", text: response }],
    };
  } catch (error) {
    return {
      content: [
        {
          type: "text",
          text: `Certificate verification failed. Certificates may not exist or are invalid.`,
        },
      ],
      isError: true,
    };
  }
}
